k-distribution/src/main/scripts/lib/pyk/pyk/kastManip.py

```diff
@@ -28,13 +31,48 @@
     mlOr,
     mlTop,
 )
-from .utils import combine_dicts, dedupe, find_common_items, hash_str
+from .utils import (
+    FrozenDict,
+    combine_dicts,
+    dedupe,
+    find_common_items,
+    hash_str,
+)
 
 T = TypeVar('T', bound=KAst)
 KI = TypeVar('KI', bound=KInner)
 W = TypeVar('W', bound=WithKAtt)
 
 
+@dataclass(frozen=True)
+class Subst(Mapping[str, KInner]):
+    subst: FrozenDict[str, KInner]
+
+    def __init__(self, subst: Mapping[str, KInner] = {}):
+        object.__setattr__(self, 'subst', FrozenDict({k: v for k, v in subst.items() if type(v) is not KVariable or v.name != k}))
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self.subst)
+
+    def __len__(self) -> int:
+        return len(self.subst)
+
+    def __getitem__(self, key: str) -> KInner:
+        return self.subst[key]
+
+    def __mul__(self, other: 'Subst') -> 'Subst':
+        return self.compose(other)
+
+    def __call__(self, term: KInner) -> KInner:
+        return self.apply(term)
+
+    def compose(self, other: 'Subst') -> 'Subst':
+        return Subst(compose_substs(self, other))
+
+    def apply(self, term: KInner) -> KInner:
+        return substitute(term, self)
+
```

Comment:
Is this a good place for this? I guess it should be with the definition of `KInner` I suppose? Should be fine for now.

Or perhaps in its own file.

Reply:
Moved to its own module.